base/geo_object_id.hpp

```diff
@@ -0,0 +1,100 @@
+//
+// The highest byte encodes one of (2^7 - 1) = 127 possible object sources.
+// Another byte is reserved and the last 6 bytes leave us with 2^48 possible values that can be
+// used for ids within a source.
+// Typically, the reserved byte will be zero but it may be used in future if the format changes.
+// At the time of writing, OSM has approximately 2^32 different nodes with ids starting
+// from one (https://wiki.openstreetmap.org/wiki/Stats) and this is by far the largest
+// serial numbers that we use.
+// The highest bit is zero so that the resulting number is positive if read as a signed 64-bit
+// integer in two's complement notation. This is important for readability in some database systems
+// that do not support unsigned integers. An older scheme we used to store OsmIds tried to keep as
+// much bits for the serial number as possible and utilized the highest bit which resulted in a lot
+// of "negative" numbers.
+//
+// When all bits are set to zero the GeoObjectId is defined to be invalid.
+//
+// Another way would be to use separate headers for source and for categories within the source,
+// as in OSM->Way->Id instead of OSMWay->Id that we have now but we do not have this many sources
+// and the difference does not seem important. Also this would probably touch the highest bit.
+class GeoObjectId
+{
+public:
+  // Sources of the objects.
+  enum class Type : uint8_t
+  {
+    Invalid = 0x00,
+    OsmNode = 0x01,
+    OsmWay = 0x02,
+    OsmRelation = 0x03,
+    BookingComNode = 0x04,
+
+    // Artificial objects that substitute objects not presented in OSM for some reason.
+    // E.g., streets that only exist in addr:street fields of houses but not as separate OSM ways.
+    OsmSurrogate = 0x05,
+
+    // Federal informational address system. http://www.ifias.ru/
+    Fias = 0x06,
+
+    ObsoleteOsmNode = 0x40,
+    ObsoleteOsmWay = 0x80,
+    ObsoleteOsmRelation = 0xC0,
+  };
+
+  static const uint64_t kInvalid = 0ULL;
+
+  explicit GeoObjectId(uint64_t encodedId = kInvalid);
+
+  GeoObjectId(Type type, uint64_t id);
+
+  // Returns the id that the object has within its source.
+  uint64_t GetSerialId() const;
+
+  // Returns the encoded value that contains both
+  // the source type and the serial number.
+  uint64_t GetEncodedId() const;
+
+  // Returns the source type of the object.
+  Type GetType() const;
+
+  bool operator<(GeoObjectId const & other) const { return m_encodedId < other.m_encodedId; }
+  bool operator==(GeoObjectId const & other) const { return m_encodedId == other.m_encodedId; }
+  bool operator!=(GeoObjectId const & other) const { return !(*this == other); }
+  bool operator==(uint64_t other) const { return GetSerialId() == other; }
+
+private:
+  uint64_t m_encodedId;
+};
+
+struct HashGeoObjectId : private std::hash<uint64_t>
+{
+  size_t operator()(GeoObjectId const & id) const
+  {
+    return std::hash<uint64_t>::operator()(id.GetSerialId());
```

Comment:
Using `GetSerialId` is invalid here

Reply:
I don't think it's invalid, but surely suboptimal.
done